src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -316,28 +313,59 @@ public boolean replace(final TransportRequest request, boolean retainMode, Strin
         }, false);
     }
 
+    private static final class IndexResolveKey {
+        private final IndicesOptions opts;
+        private final boolean enableCrossClusterResolution;
+        private final String[] original;
+        public IndexResolveKey(IndicesOptions opts, boolean isSearchOrFieldCapabilities, String[] original) {
+            this.opts = opts;
+            this.enableCrossClusterResolution = isSearchOrFieldCapabilities;
+            this.original = original;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            IndexResolveKey that = (IndexResolveKey) o;
+            return enableCrossClusterResolution == that.enableCrossClusterResolution &&
+                    opts.equals(that.opts) &&
+                    Arrays.equals(original, that.original);
+        }
+
+        @Override
+        public int hashCode() {
+            return Boolean.hashCode(enableCrossClusterResolution) + 31 * opts.hashCode() + 127*Arrays.hashCode(original);
+        }
+    }
+
     public Resolved resolveRequest(final Object request) {
         if(log.isDebugEnabled()) {
             log.debug("Resolve aliases, indices and types from {}", request.getClass().getSimpleName());
         }
 
         final Resolved.Builder resolvedBuilder = new Resolved.Builder();
         final AtomicBoolean isIndicesRequest = new AtomicBoolean();
-        getOrReplaceAllIndices(request, new IndicesProvider() {
-
-            @Override
-            public String[] provide(String[] original, Object localRequest, boolean supportsReplace) {
-                final IndicesOptions indicesOptions = indicesOptionsFrom(localRequest);
-                final Resolved iResolved = resolveIndexPatterns(indicesOptions, localRequest, original);
+        // set of previously resolved index requests to avoid resolving
+        // the same index more than once while processing bulk requests
+        final Set<IndexResolveKey> alreadyResolved = new HashSet<>();
+
+        getOrReplaceAllIndices(request, (original, localRequest, supportsReplace) -> {
+            final IndicesOptions indicesOptions = indicesOptionsFrom(localRequest);
+            final boolean isSearchOrFieldCapabilities = localRequest instanceof FieldCapabilitiesRequest || localRequest instanceof SearchRequest;
```

Comment:
Either inline or rename to `enableCrossClusterResolution`.

Reply:
Yeah, missed that one - fixed!